### ISSUE TITLE
ability to define re-usable code libraries and import them where needed

### ISSUE THREAD

**Issue by kmahmood74** [EnsembleUI/ensemble#62]
1. Defining re-usable code
Just like we define screens and widgets as re-usable, code itself is defined as another artifact except this artifact contains javascript code. 
Re-usable code contains - 
- functions such as
```js
function add(a,b) { return a+b;}
```
- vars such as 
```js
var counter = 0;
```
2. Scope of imports
Each import is resolved at the screen level. Code libraries cannot be used to share state across screens or widgets. 
Example
myLib
```js
var counter = 0;
```
Let's say myLib is imported in `Home` and `Profile` screens. The code in the `Home` screen updates the `counter` to 10. The `Profile` screen or any other screen that imported myLib will still get their own independent `counter` as 0. 

3. How are the variables and functions in re-usable libraries resolved in case of naming conflicts
We apply the rules that ES5/HTML uses i.e. the last one wins except when the same library has been included multiple times. 
Example - 
commonUtils
```js
var counter = 0;
function add(a,b) {return a + b;}
```
apiUtils
```js
var counter = 10;
function add(a,b,c) {return a+b+c;}
```

Let's say the screen imports commonUtils and then apiUtils. Value of the counter will be 10 and the add function from apiUtils will be one invoked when anyone calls add in the page. 

Same applies to re-usable widgets as well. Code libraries at the screen level are automatically inherited by all the widgets in the page. Widgets can have their own libraries as well. If widget imports the same library that is already imported in the screen it resides in, it is *not* imported again. Any new libraries are imported in the order defined.

5. Importing code is supported in both Screens and Widgets as follows. Currently importing code within a code library is not supported. It is being tracked here - #1142 

Importing Code in a screen - 

```
Import:
  - apiUtils
  - common

View:
    # Optional - style the screen
    styles:
      scrollableView: true

    # Optional - set the header for the screen
    header:
      title: Home
    onLoad: |
      saveName('Khurram','Mahmood');
      console.log('name is ='+storageName);

    # Specify the body of the screen
    body:
      Column:
        styles:
          padding: 24
          gap: 8
        children:
          - Text:
              text: Hello ${ensemble.storage.jslibtest.name.first} ${ensemble.storage.jslibtest.name.last}
```
Importing in a re-usable widget - 

```
Import:
  - common
  - apiUtils
Widget:
    inputs:
      - inputName
    body:
      Column:
        children:
          - Text:
              text: ${inputName.first} ${inputName.last}
          - Button:
              text: ${storageName.name.first} ${storageName.name.last}
              onTap: |-
                sayHello();
                console.log('name:'+getName());
                callAPI('mockAPI',{});
```

**Comment by kmahmood74** [comment]
to be considered but not urgent

**Comment by kmahmood74** [comment]
closing as stale

**Comment by amin-nas** [comment]
@kmahmood74 I'm testing  this locally but I get the following error:
![image](https://github.com/EnsembleUI/ensemble/assets/491202/074344db-5e0f-4ff4-8d2d-bc2eec0d6e70)

Can you add a simple example to Starter repo?

**Comment by kmahmood74** [comment]
@amin-nas it won't work until GCP function changes are made. Please wait for https://github.com/EnsembleUI/ensemble-web-studio/issues/683 to be completed. The GCP function has to be modified to take the code and combine it into one resource file.

**Comment by amin-nas** [comment]
@kmahmood74 I'm asking from perspective of someone not using Studio, and using their own editor. See widgets example here for that use case. We need an example similarly for JS codes.

https://github.com/EnsembleUI/starter/blob/main/ensemble/apps/helloApp/resources.ensemble

**Comment by kmahmood74** [comment]
> @kmahmood74 I'm asking from perspective of someone not using Studio, and using their own editor. See widgets example here for that use case. We need an example similarly for JS codes.
> 
> https://github.com/EnsembleUI/starter/blob/main/ensemble/apps/helloApp/resources.ensemble

That's exactly right. The same resources.ensemble will have the js code as well. See the sample yaml in https://github.com/EnsembleUI/ensemble-web-studio/issues/683, it's all inside resources.ensemble

When someone is running ensemble in their own IDE and not in studio, this whole thing about combining widgets and other internal resources in one file won't work if the user wants to change the code or the re-usable widget as it is done today in a GCP function. This is designed to for only the ensemble studio dev environment, it could be downloaded and deployed but not developed locally. This is a design choice that @vusters made

**Comment by kmahmood74** [comment]
I'll add a local example for code as well. But remember no one can use it as a dev thing, it's only when you have downloaded from ensemble and are going to deploy it. 

A big departure we made when optimizing ensemble custom widget is that we started processing the YAML and storing a derivative instead of just storing the raw yaml. This is not an optimal design decision as it negates everything from - people checking in manually into their githubs and deploying from there or developing locally. 

We need to allow internal resources as raw yaml when people are developing locally. Right now it is not right
